File: knowledgebase-sdk/src/main/java/ru/usedesk/knowledgebase_sdk/internal/domain/KnowledgeBaseInteractor.java
package ru.usedesk.knowledgebase_sdk.internal.domain;


import androidx.annotation.NonNull;

import java.util.List;

import javax.inject.Inject;
import javax.inject.Named;

import io.reactivex.Completable;
import io.reactivex.Scheduler;
import io.reactivex.Single;
import io.reactivex.SingleEmitter;
import io.reactivex.SingleOnSubscribe;
import ru.usedesk.common_sdk.external.entity.exceptions.UsedeskException;
import ru.usedesk.knowledgebase_sdk.external.IUsedeskKnowledgeBase;
import ru.usedesk.knowledgebase_sdk.external.entity.UsedeskArticleBody;
import ru.usedesk.knowledgebase_sdk.external.entity.UsedeskArticleInfo;
import ru.usedesk.knowledgebase_sdk.external.entity.UsedeskCategory;
import ru.usedesk.knowledgebase_sdk.external.entity.UsedeskKnowledgeBaseConfiguration;
import ru.usedesk.knowledgebase_sdk.external.entity.UsedeskSearchQuery;
import ru.usedesk.knowledgebase_sdk.external.entity.UsedeskSection;
import ru.usedesk.knowledgebase_sdk.internal.data.repository.IKnowledgeBaseRepository;

public class KnowledgeBaseInteractor implements IUsedeskKnowledgeBase {

    private final IKnowledgeBaseRepository knowledgeRepository;
    private final Scheduler workScheduler;
    private final Scheduler mainThreadScheduler;
    private final UsedeskKnowledgeBaseConfiguration configuration;

    @Inject
    KnowledgeBaseInteractor(@NonNull IKnowledgeBaseRepository knowledgeRepository,
                            @NonNull @Named("work") Scheduler workScheduler,
                            @NonNull @Named("main") Scheduler mainThreadScheduler,
                            @NonNull UsedeskKnowledgeBaseConfiguration configuration) {
        this.knowledgeRepository = knowledgeRepository;
        this.workScheduler = workScheduler;
        this.mainThreadScheduler = mainThreadScheduler;
        this.configuration = configuration;
    }

    private <T> Single<T> createSingle(SingleOnSubscribe<T> emitter) {
        return Single.create(new SafeSingleEmitter<>(emitter))
                .subscribeOn(workScheduler)
                .observeOn(mainThreadScheduler);
    }

    @Override
    @NonNull
    public Single<List<UsedeskSection>> getSectionsRx() {
        return createSingle(emitter -> emitter.onSuccess(getSections()));
    }

    @Override
    @NonNull
    public Single<UsedeskArticleBody> getArticleRx(long articleId) {
        return createSingle(emitter -> emitter.onSuccess(getArticle(articleId)));
    }

    @NonNull
    @Override
    public Single<List<UsedeskCategory>> getCategoriesRx(long sectionId) {
        return createSingle(emitter -> emitter.onSuccess(getCategories(sectionId)));
    }

    @NonNull
    @Override
    public Single<List<UsedeskArticleInfo>> getArticlesRx(long categoryId) {
        return createSingle(emitter -> emitter.onSuccess(getArticles(categoryId)));
    }

    @Override
    @NonNull
    public Single<List<UsedeskArticleBody>> getArticlesRx(@NonNull String searchQuery) {
        return createSingle(emitter -> emitter.onSuccess(getArticles(searchQuery)));
    }

    @Override
    @NonNull
    public Single<List<UsedeskArticleBody>> getArticlesRx(@NonNull UsedeskSearchQuery searchQuery) {
        return createSingle(emitter -> emitter.onSuccess(getArticles(searchQuery)));
    }

    @Override
    @NonNull
    public Completable addViewsRx(long articleId) {
        return Completable.create(emitter -> {
            addViews(articleId);
            emitter.onComplete();
        }).subscribeOn(workScheduler)
                .observeOn(mainThreadScheduler);
    }

    @Override
    public void addViews(long articleId) throws UsedeskException {
        knowledgeRepository.addViews(configuration.getAccountId(), configuration.getToken(), articleId);
    }

    @Override
    @NonNull
    public List<UsedeskCategory> getCategories(long sectionId) throws UsedeskException {
        return knowledgeRepository.getCategories(configuration.getAccountId(), configuration.getToken(), sectionId);
    }

    @Override
    @NonNull
    public List<UsedeskSection> getSections() throws UsedeskException {
        return knowledgeRepository.getSections(configuration.getAccountId(), configuration.getToken());
    }

    @Override
    @NonNull
    public UsedeskArticleBody getArticle(long articleId) throws UsedeskException {
        return knowledgeRepository.getArticleBody(configuration.getAccountId(), configuration.getToken(), articleId);
    }

    @Override
    @NonNull
    public List<UsedeskArticleBody> getArticles(@NonNull String searchQuery) throws UsedeskException {
        UsedeskSearchQuery query = new UsedeskSearchQuery.Builder(searchQuery).build();

        return knowledgeRepository.getArticles(configuration.getAccountId(), configuration.getToken(), query);
    }

    @Override
    @NonNull
    public List<UsedeskArticleBody> getArticles(@NonNull UsedeskSearchQuery searchQuery) throws UsedeskException {
        return knowledgeRepository.getArticles(configuration.getAccountId(), configuration.getToken(), searchQuery);
    }

    @Override
    @NonNull
    public List<UsedeskArticleInfo> getArticles(long categoryId) throws UsedeskException {
        return knowledgeRepository.getArticles(configuration.getAccountId(), configuration.getToken(), categoryId);
    }

    static class SafeSingleEmitter<T> implements SingleOnSubscribe<T> {
        private final SingleOnSubscribe<T> singleOnSubscribeSafe;

        SafeSingleEmitter(SingleOnSubscribe<T> singleOnSubscribeSafe) {
            this.singleOnSubscribeSafe = singleOnSubscribeSafe;
        }

        @Override
        public void subscribe(@NonNull SingleEmitter<T> emitter) throws Exception {
            try {
                singleOnSubscribeSafe.subscribe(emitter);
            } catch (Exception e) {
                if (!emitter.isDisposed()) {
                    throw e;
                }
            }
        }
    }
}

File: knowledgebase-sdk/src/main/java/ru/usedesk/knowledgebase_sdk/internal/domain/KnowledgeBaseInteractor.kt
package ru.usedesk.knowledgebase_sdk.internal.domain

import io.reactivex.*
import ru.usedesk.common_sdk.external.entity.exceptions.UsedeskException
import ru.usedesk.knowledgebase_sdk.external.IUsedeskKnowledgeBase
import ru.usedesk.knowledgebase_sdk.external.entity.*
import ru.usedesk.knowledgebase_sdk.internal.data.repository.IKnowledgeBaseRepository
import toothpick.InjectConstructor
import javax.inject.Named

@InjectConstructor
class KnowledgeBaseInteractor(
        private val knowledgeRepository: IKnowledgeBaseRepository,
        @Named("work")
        private val workScheduler: Scheduler,
        @Named("main")
        private val mainThreadScheduler: Scheduler,
        private val configuration: UsedeskKnowledgeBaseConfiguration
) : IUsedeskKnowledgeBase {

    private fun <T> createSingle(emitter: SingleOnSubscribe<T>): Single<T> {
        return Single.create(SafeSingleEmitter(emitter))
                .subscribeOn(workScheduler)
                .observeOn(mainThreadScheduler)
    }

    override fun getSectionsRx(): Single<List<UsedeskSection>> {
        return createSingle {
            it.onSuccess(getSections())
        }
    }

    override fun getArticleRx(articleId: Long): Single<UsedeskArticleBody> {
        return createSingle {
            it.onSuccess(getArticle(articleId))
        }
    }

    override fun getCategoriesRx(sectionId: Long): Single<List<UsedeskCategory>> {
        return createSingle {
            it.onSuccess(getCategories(sectionId))
        }
    }

    override fun getArticlesRx(categoryId: Long): Single<List<UsedeskArticleInfo>> {
        return createSingle {
            it.onSuccess(getArticles(categoryId))
        }
    }

    override fun getArticlesRx(searchQuery: String): Single<List<UsedeskArticleBody>> {
        return createSingle {
            it.onSuccess(getArticles(searchQuery))
        }
    }

    override fun getArticlesRx(searchQuery: UsedeskSearchQuery): Single<List<UsedeskArticleBody>> {
        return createSingle {
            it.onSuccess(getArticles(searchQuery))
        }
    }

    override fun addViewsRx(articleId: Long): Completable {
        return Completable.create { emitter: CompletableEmitter ->
            addViews(articleId)
            emitter.onComplete()
        }.subscribeOn(workScheduler)
                .observeOn(mainThreadScheduler)
    }

    @Throws(UsedeskException::class)
    override fun addViews(articleId: Long) {
        knowledgeRepository.addViews(configuration.accountId, configuration.token, articleId)
    }

    @Throws(UsedeskException::class)
    override fun getCategories(sectionId: Long): List<UsedeskCategory> {
        return knowledgeRepository.getCategories(configuration.accountId, configuration.token, sectionId)
    }

    @Throws(UsedeskException::class)
    override fun getSections(): List<UsedeskSection> {
        return knowledgeRepository.getSections(configuration.accountId, configuration.token)
    }

    @Throws(UsedeskException::class)
    override fun getArticle(articleId: Long): UsedeskArticleBody {
        return knowledgeRepository.getArticleBody(configuration.accountId, configuration.token, articleId)
    }

    @Throws(UsedeskException::class)
    override fun getArticles(searchQuery: String): List<UsedeskArticleBody> {
        val query = UsedeskSearchQuery.Builder(searchQuery).build()
        return knowledgeRepository.getArticles(configuration.accountId, configuration.token, query)
    }

    @Throws(UsedeskException::class)
    override fun getArticles(searchQuery: UsedeskSearchQuery): List<UsedeskArticleBody> {
        return knowledgeRepository.getArticles(configuration.accountId, configuration.token, searchQuery)
    }

    @Throws(UsedeskException::class)
    override fun getArticles(categoryId: Long): List<UsedeskArticleInfo> {
        return knowledgeRepository.getArticles(configuration.accountId, configuration.token, categoryId)
    }

    internal class SafeSingleEmitter<T>(
            private val singleOnSubscribeSafe: SingleOnSubscribe<T>
    ) : SingleOnSubscribe<T> {

        @Throws(Exception::class)
        override fun subscribe(emitter: SingleEmitter<T>) {
            try {
                singleOnSubscribeSafe.subscribe(emitter)
            } catch (e: Exception) {
                if (!emitter.isDisposed) {
                    throw e
                }
            }
        }
    }
}